common/init.go

```diff
@@ -0,0 +1,98 @@
+package common
+
+import (
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/getlantern/radiance/app"
+	"github.com/getlantern/radiance/common/reporting"
+	"github.com/getlantern/radiance/internal"
+)
+
+const (
+	// envLogLevel is the environment variable that can be used to override the log level.
+	envLogLevel     = "RADIANCE_LOG_LEVEL"
+	defaultLogLevel = "info"
+)
+
+var (
+	initMutex   sync.Mutex
+	initialized bool
+)
+
+// Init initializes the common components of the application. This includes setting up the directories
+// for data and logs, initializing the logger, and setting up reporting.
+func Init(dataDir, logDir, logLevel string) error {
+	initMutex.Lock()
+	defer initMutex.Unlock()
+	if initialized {
+		return nil
+	}
+
+	reporting.Init(Version)
+	dataDir, logDir, err := SetupDirectories(dataDir, logDir)
+	if err != nil {
+		return fmt.Errorf("failed to setup directories: %w", err)
+	}
+
+	err = initLogger(filepath.Join(logDir, app.LogFileName), logLevel)
+	if err != nil {
+		return fmt.Errorf("initialize log: %w", err)
+	}
+	initialized = true
+	return nil
+}
+
+// initLogger reconfigures the default slog.Logger to write to a file and stdout and sets the log level.
+// The log level is determined, first by the environment variable if set and valid, then by the provided level.
+// If both are invalid and/or not set, it defaults to "info".
+func initLogger(logPath, level string) error {
+	var lvl slog.Level
+	var err error
+	envLevelValue := os.Getenv(envLogLevel)
+	logLevelSet := false
+	if envLevelValue != "" {
+		if lvl, err = internal.ParseLogLevel(envLevelValue); err != nil {
+			slog.Warn("Failed to parse "+envLogLevel, "error", err)
+		} else {
+			logLevelSet = true
+		}
+	}
+	if !logLevelSet && level != "" {
+		if lvl, err = internal.ParseLogLevel(level); err != nil {
+			slog.Warn("Failed to parse given log level", "error", err)
+		}
+	}
+	slog.SetLogLoggerLevel(lvl)
+
+	// If the log file does not exist, create it.
+	f, err := os.OpenFile(logPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
+	if err != nil {
+		return fmt.Errorf("failed to open log file: %w", err)
+	}
+	// defer f.Close() - file should be closed externally when logger is no longer needed
```

Comment:
In this case, we need to make Close function or make it part of radiance shutdown process